src/storage/mutate/UpdateVertexProcessor.cpp

```diff
@@ -259,13 +261,41 @@ std::string UpdateVertexProcessor::updateAndWriteBack() {
         }
     }
 
-    std::vector<kvstore::KV> data;
+    std::unique_ptr<kvstore::BatchHolder> batchHolder = std::make_unique<kvstore::BatchHolder>();
     for (const auto& u : tagUpdaters_) {
-        data.emplace_back(std::move(u.second->key),
-                          u.second->updater->encode());
+        auto nKey = u.second->kv.first;
+        auto nVal = u.second->updater->encode();
+        batchHolder->put(std::move(nKey), std::move(nVal));
+        if (!indexes_.empty()) {
+            for (auto &index : indexes_) {
+                if (index.get_tagOrEdge() == u.first) {
+                    auto prop = u.second->updater->encode();
+                    auto reader = RowReader::getTagPropReader(this->schemaMan_,
+                                                              std::move(prop),
+                                                              spaceId_,
+                                                              u.first);
+                    const auto &cols = index.get_cols();
+                    auto values = collectIndexValues(std::move(reader).get(), cols);
```

Comment:
why std::move(reader) ?

Reply:
Good point! It got me thinking about performance optimization. I've solved the performance problem for delete|update vertices or edges.

Reply:
> why std::move(reader) ?

Good point! It got me thinking about performance optimization. I've solved the performance problem for delete|update vertices or edges.